### PR TITLE
CloudMonitor: Iterate over pageToken for resources

### LINKED ISSUE #36206: [Google Cloud Monitoring] Query builder trims service/metric list over 10000
**What happened**: When trying to build a query using the Google Cloud Monitoring Plugin, I can't  find or select some of the GCP Metric services to show the metrics:
image

Last service shown is `Kubernetes` and in this service the last metric I can select is `grpc_server_handling_seconds`:
image

Looking to the requests seems to me that the request to `/api/datasources/proxy/9/cloudmonitoring/v3/projects/{project}/metricDescriptors` is being trimmed. Checking how many metrics are returned by this endpoints turns out that they are 10000:
```
# curl -s 'https://org.grafana.net/api/datasources/proxy/9/cloudmonitoring/v3/projects/{project}/metricDescriptors' -H <HEADERS> --compressed | jq . | grep '"name": "projects/{project}/metricDescriptors' | wc -l
 10000
 ```

So I cannot use the query builder to build queries from the services that are trimmed out form the `metricDescriptors` endpoint.

**What you expected to happen**: The building allows me to select all the GCP services/metrics available in the project.

**How to reproduce it (as minimally and precisely as possible)**: Try to build a query from a GCP project with more than 10000 metric descriptors.

**Anything else we need to know?**: Using grafana cloud.

**Environment**:
- Grafana version: v8.0.3 (cae5c5e46b) (grafana cloud)
- Data source type & version: 1.0.0
- OS Grafana is installed on: Grafana cloud
- User OS & Browser: Tested with Linux/MacOS and Firefox/Chrome.
- Grafana plugins: Google Cloud Monitoring
- Others:

### PR DESCRIPTION
**Pull request by iwysiu**
**What this PR does / why we need it**:
Pages through the cloud monitoring resources before returning them.

**Which issue(s) this PR fixes**:

Fixes #36206 

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/tsdb/cloudmonitoring/resource_handler.go [hunk 1]
```diff
@@ -235,10 +235,57 @@ func processData(res *http.Response, responseFn processResponse) ([]byte, int, e
 	if err != nil {
 		return nil, http.StatusInternalServerError, fmt.Errorf("unable to encode response %v", err)
 	}
-
 	return buf.Bytes(), 0, nil
 }
 
+func processData(req *http.Request, cli *http.Client, responseFn processResponse) ([]byte, http.Header, int, error) {
```
Comment (andresmgot) [inline pkg/tsdb/cloudmonitoring/resource_handler.go:241]: didn't spend too much time looking at this but this function seems to do too many things. Maybe we can split it to different ones that can be easier to test? (if needed). From a first glance, it seems that there is code to do a loop over requests, parse the result and then prepare to send the response. Maybe that's a good subdivision?

Also, `processData` does not match what this function does anymore so it requires some naming, easier if you split it :)

**Review by andresmgot — COMMENTED** [review]
I didn't have time to do a full review, will do it tomorrow but let me know if my suggestion makes sense!

File: pkg/tsdb/cloudmonitoring/resource_handler.go (hunk 1 above)
Comment (iwysiu) [reply pkg/tsdb/cloudmonitoring/resource_handler.go]: I refactored a bit, subdividing it is a bit weird, since there are a lot of variables to pass

**Review by andresmgot — COMMENTED** [review]
I like the changes! More chewable now :)

Mostly LGTM, I have some minor comments.

File: pkg/tsdb/cloudmonitoring/resource_handler.go [hunk 2]
```diff
@@ -235,10 +241,67 @@ func processData(res *http.Response, responseFn processResponse) ([]byte, int, e
...
+
+		var errcode int
+		responses, token, errcode, err = processData(res.Body, encoding, responses, responseFn)
+		if err != nil {
+			return nil, nil, "", errcode, err
+		}
+
+		if token == "" {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/tsdb/cloudmonitoring/resource_handler.go]: done. I think it's theoretically possible for different api calls to encode differently, so I'm still checking that.

File: pkg/tsdb/cloudmonitoring/resource_handler.go [hunk 3]
```diff
@@ -233,10 +243,74 @@ func processData(res *http.Response, responseFn processResponse) ([]byte, int, e
...
+
+	responses, token, errcode, err := processData(res.Body, encoding, responseFn)
+	if err != nil {
+		code = errcode
+	}
+	return encoding, originalHeader, responses, token, code, err
+}
+
+func getResponses(req *http.Request, cli *http.Client, responseFn processResponse) ([]json.RawMessage, http.Header, string, int, error) {
+	encoding, originalHeader, responses, token, originalCode, err := doRequest(req, cli, responseFn)
```
Comment (andresmgot) [inline pkg/tsdb/cloudmonitoring/resource_handler.go:286]: nit: what do you think about having a struct with these parameters? that would be easier to update and extract data from. Something like `multipartResponse`

**Review by andresmgot — APPROVED** [review]
LGTM. Thanks for the changes!

**Review by andresmgot — APPROVED** [review]

**Comment by andresmgot** [comment]
please, add the backport/v8.3.x label and the milestone so this is included in the next release! I'd also add the label "add to changelog" to notify users.

**Pull request merged**